### DISCUSSION TITLE
Store generic ActorRef

### DISCUSSION THREAD

**Discussion by rubengrim** [tqwewe/kameo#38] [Q&A] [answered]
I'm trying to store an `ActorRef` of a generic actor that implements a specific message, but can't get it to work. Been reading through some of the source code and tried something like this: 

```rust
#[derive(Actor)]
pub struct SomeActor<A>
where
    A: Actor + Message<SomeMessage>,
    ActorRef<A>: Request<A, SomeMessage, A::Mailbox>,
{
    other_actor: ActorRef<A>,
}

// Send SomeMessage like this later
self.other_actor.ask(SomeMessage).send().await;
```

But I get this error when sending:
``no method named `send` found for struct `AskRequest<A, <A as Actor>::Mailbox, ReadyForNext, WithoutRequestTimeout, WithoutRequestTimeout>` in the current scope``

The `PubSub` implementation seems to use a wrapper trait for the subscribers and I'm thinking that I might need to do something similar, but I honestly don't understand it. Any help in the right direction is appreciated!

**Comment by tqwewe** [comment]
Hi rubengrim, this is perfect timing for something I've been working on today actually! I just saw this question are creating a PR with my rework of how sending is done. https://github.com/tqwewe/kameo/pull/39

With these new changes, all you should need are the following bounds for ask requests:
https://github.com/tqwewe/kameo/blob/ad1984a133963f60ddf44a6efe19902a4d3c5222/src/request/ask.rs#L166-L169
In your case, since you're directly using `SomeMessage`, you would just need:
```rust
#[derive(Actor)]
pub struct SomeActor<A>
where
    A: Actor<Mailbox = BoundedMailbox<A>> + Message<SomeMessage>,
{
    other_actor: ActorRef<A>,
}
```

However, this approach requires the mailbox to be known. If you want a super generic approach, you could do something like this:
```rust
#[derive(Actor)]
struct SomeActor<A, Tm, Tr>
where
    A: Actor + Message<SomeMessage>,
    AskRequest<
        LocalAskRequest<A, A::Mailbox>,
        A::Mailbox,
        SomeMessage,
        Tm,
        Tr,
    >: MessageSend,
{
    other_actor: ActorRef<A>,
}
```
`Tm` and `Tr` here are generics for whether the ask request has a mailbox timeout or reply timeout set.
Eg:
```rust
actor_ref.ask(...).mailbox_timeout(...).reply_timeout(...).send().await
```
If you know there will or will not be timeouts used here, you can hard code them to be `WithRequestTimeout` or `WithoutRequestTimeout`.

An alternative to above with no timeouts available:
```rust
#[derive(Actor)]
struct SomeActor<A>
where
    A: Actor + Message<SomeMessage>,
    AskRequest<
        LocalAskRequest<A, A::Mailbox>,
        A::Mailbox,
        SomeMessage,
        WithoutRequestTimeout,
        WithoutRequestTimeout,
    >: MessageSend,
{
    other_actor: ActorRef<A>,
}
```

It's quite verbose, but is pretty powerful for allowing different ways of sending messages based on mailbox type, or if there's timeouts used.

**Reply by rubengrim** [reply]
Yeah that's great timing! Mailbox/timeout will be known so the first example will work great. Thank you for the help and nice work on the crate, really enjoying using it!

**Reply by tqwewe** [reply, marked as answer]
Very cool of you to be looking into the source code :)

I have realised that your requirements aren't really dependant on the new PR I mentioned. I believe all you needed was to specify the mailbox there too. `A: Actor<Mailbox = BoundedMailbox<A>> + Message<SomeMessage>`

The bounds required based on the current version of kameo are here:
https://github.com/tqwewe/kameo/blob/6f7b6f76b6043be426948da0a9621ed4e82018db/src/request/ask.rs#L374-L389

**Reply by rubengrim** [reply]
Works perfectly, thanks again!

**Comment by tqwewe** [comment]
Just posting an update to this, with the PR https://github.com/tqwewe/kameo/pull/71, this issue should never come up anymore, as most request types work regardless of the actors mailbox.
